mule-migration-tool-library/src/main/java/com/mulesoft/tools/migration/library/mule/steps/salesforce/QueryOperation.java

```diff
@@ -0,0 +1,140 @@
+import com.mulesoft.tools.migration.util.ExpressionMigrator;
+import org.jdom2.CDATA;
+import org.jdom2.Element;
+
+import java.util.List;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+/**
+ * Migrate Query operation
+ *
+ * @author Mulesoft Inc.
+ * @since 1.0.0
+ */
+public class QueryOperation extends AbstractApplicationModelMigrationStep implements ExpressionMigratorAware {
+
+  private static final String name = "query";
+
+  private ExpressionMigrator expressionMigrator;
+
+  public QueryOperation() {
+    this.setAppliedTo(XmlDslUtils.getXPathSelector(SalesforceConstants.MULE3_SALESFORCE_NAMESPACE_URI, name, false));
+    this.setNamespacesContributions(newArrayList(SalesforceConstants.MULE3_SALESFORCE_NAMESPACE));
+  }
+
+  @Override
+  public void execute(Element mule3QueryOperation, MigrationReport report) throws RuntimeException {
+    getApplicationModel().addNameSpace(SalesforceConstants.MULE4_SALESFORCE_NAMESPACE,
+                                       SalesforceConstants.MULE4_SALESFORCE_NAMESPACE_URI, mule3QueryOperation.getDocument());
+
+    Element mule4QueryOperation = new Element(name, SalesforceConstants.MULE4_SALESFORCE_NAMESPACE);
+    setAttributes(mule3QueryOperation, mule4QueryOperation);
+
+    if (mule3QueryOperation.getAttribute("accessTokenId") != null) {
+      report.report("salesforce.accessTokenId", mule3QueryOperation, mule4QueryOperation);
+    }
+
+    String query = mule3QueryOperation.getAttributeValue("query");
+    if (query != null) {
+      Element mule4Query = new Element("salesforce-query", SalesforceConstants.MULE4_SALESFORCE_NAMESPACE);
+      String expression = expressionMigrator.migrateExpression(query, true, mule3QueryOperation);
+      mule4Query.setContent(new CDATA(expression));
+
+      mule4QueryOperation.addContent(mule4Query);
+    }
+
+    Element mule3Headers = mule3QueryOperation.getChild("headers", SalesforceConstants.MULE3_SALESFORCE_NAMESPACE);
+    String fetchSize = mule3QueryOperation.getAttributeValue("fetchSize");
+    report.report("salesforce.fetchSize", mule3QueryOperation, mule4QueryOperation);
+    if (mule3Headers != null) {
+      String refHeaders = mule3Headers.getAttributeValue("ref");
+
+      if (refHeaders != null) {
+        String expression = expressionMigrator.migrateExpression(refHeaders, true, mule3Headers);
+
+        String unwrappedExpression = expressionMigrator.unwrap(expression);
+        if (fetchSize != null) {
+          expression = expressionMigrator.wrap(unwrappedExpression + " ++ {'batchSize':\"" + fetchSize + "\"}");
+        } else {
+          expression = expressionMigrator.wrap(unwrappedExpression + " ++ {'batchSize':\"2000\"}");
+        }
+
+        mule4QueryOperation.setAttribute("headers", expression);
+      }
+
+      List<Element> children = mule3Headers.getChildren();
+      if (children.size() > 0) {
+        Element mule4Headers = new Element("headers", SalesforceConstants.MULE4_SALESFORCE_NAMESPACE);
```

Comment:
I'm seeing this code duplicated on almost all the operations. It will be better to extract it to an abstract class and reuse it on all the operations. Because if we need to change something there, we will need to look anywhere this logic it's defined.
Maybe you could create an AbstractSalesforceApplicationModelMigrationStep class and define the method there. Then consume it on all the operations.

Reply:
Yes, we saw that. We have a task for the next sprint to get rid of all this duplicate code

Reply:
Great! :+1: